streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -267,7 +283,17 @@ public void close() {
 
     private void checkForException() {
         if (sendException != null) {
-            throw sendException;
+            if (sendException.getCause() instanceof KafkaException
+                && sendException.getCause().getMessage().equals("Failing batch since transaction was aborted")) {
```

Comment:
Should we avoid passing this exception to ```ProductionExceptionHandler``` as it never breaks sent now.

Reply:
Makes sense, I think we can avoid `checkForException` in `closeDirty`, and if that is sufficient we can remove the exception handling based on error message.

Reply:
That sounds right, we should never see this exception outside of `closeDirty` since we should then close all tasks dirty if the transaction is aborted.
But as for whether to check it in `closeDirty`, I think we would need to at the very least check it so we can reset the exception afterwards. Or do you think it's "safe" to just blindly reset the exception in the case of a dirty close, no matter what it was?

Reply:
1. If we are calling from `closeDirty`, then exceptions from `recordCollector.close` would be swallowed and logged.

2. If we are calling from `closeAndRecycleState`, then the upper caller from TaskManager would capture all RE and move the task to toCloseDirty.

3. The only thing that needs to be careful is calling from `closeClean` as Matthias pointed as an example above.

So looking from that side, to keep the error logged down maybe we should still always checkForException in `RecordCollector.close` while at the same time keep this hacky error message handling inside RC.

Reply:
I _think_, that we should never hit this exception during `closeClean`, that is, we should never call `closeClean` on a task/record collector that was part of an aborted transaction. 

The transaction will be aborted if a task had to be closed dirty due to:
1. suspend/preCommit/commit failed: in this. case, any task that has an ongoing transaction will throw an exception at the beginning of `closeClean` and then be closed dirty
2. postCommit/closeClean fails: in this case, one task might throw during `postCommit` or `closeClean` and cause it to be closed dirty while all other tasks are closed clean. But by definition, we have just committed the transaction, so there is no in-flight txn to be aborted

So, I think it is ok to just do this special exception handling inside `closeDirty`, and reset the `sendException` to `null` if we find this exception. WDYT?

Reply:
> I think, that we should never hit this exception during closeClean, that is, we should never call closeClean on a task/record collector that was part of an aborted transaction.

Agreed.

I was hoping we don't need this hack at all? Why do we want to call `checkForException` in close dirty?

Reply:
Well, to still log the error message while swallowing it :)

Reply:
Ok, we can do that, we only get the first exception back anyway. -- And for this, we also don't need the "hack" to check for a specific exception. In `closeDirty()` we just do
```
try {
  checkForException():
} catch (final RuntimeException logAndSwallow) {
   log.error(...);
}
```

Reply:
If we revive a task, we don't recreate the record collector AFAICT. So there may still be a `sendException` hanging around even after we `close` the record collector. If this was a truly-fatal exception, we'll check and throw it. But we shouldn't rethrow this particular non-fatal exception. Therefore, we need to check for it and reset the `sendException` iff we find this exact exception

Reply:
What about clearing the exception in `checkForException` instead:
```
private void checkForException() {
    if (sendException != null) {
        final KafkaException rethrow = sendException;
        sendException = null;
        throw rethrow;
    }
}
```
We need to fix `sendException` to make it thread safe though.